lib/pdcleaner/pdcleaner.go

```diff
@@ -218,33 +221,89 @@ func (p *pdcleaner) CleanOnce() error {
+		}
 
 		for _, d := range completeDirectDeals {
-			// AllocationID and ClaimID should match
 			cID := verifregtypes.ClaimId(d.AllocationID)
 			c, ok := claims[cID]
+			// If claim found
 			if ok {
-				present, err := activeSectors.IsSet(uint64(c.Sector))
+				// Claim Sector number and Deal Sector number should match(regardless of how DDO works)
+				// If they don't match and older sector is removed, then we can't use the metadata
+				// This check can be removed once Curio has resealing enabled, and it can provide
+				// new replacement sector details to Boost before deal reached "Complete" state.
+				if c.Sector != d.SectorID {
+					err = p.pd.RemoveDealForPiece(p.ctx, d.PieceCID, d.ID.String())
+					if err != nil {
+						// Don't return if cleaning up a deal results in error. Try them all.
+						log.Errorf("cleaning up direct deal %s for piece %s: %s", d.ID.String(), d.PieceCID, err.Error())
+					}
+					continue
+				}
+				present, err := finalSectors.IsSet(uint64(c.Sector))
 				if err != nil {
 					return fmt.Errorf("checking if bitfield is set: %w", err)
 				}
 				// Each claim is created with ProveCommit message. So, a sector in claim cannot be unproven.
 				// it must be either Active(Proving, Faulty, Recovering) or terminated. If bitfield is not set
 				// then sector must have been terminated. This method will also account for future change in sector numbers
 				// of a claim. Even if the sector is changed then it must be Active as this change will require a
-				// ProveCommit message
+				// ProveCommit message.
 				if !present {
 					err = p.pd.RemoveDealForPiece(p.ctx, d.PieceCID, d.ID.String())
 					if err != nil {
 						// Don't return if cleaning up a deal results in error. Try them all.
-						log.Errorf("cleaning up legacy deal %s for piece %s: %s", d.ID.String(), d.PieceCID, err.Error())
+						log.Errorf("cleaning up direct deal %s for piece %s: %s", d.ID.String(), d.PieceCID, err.Error())
 					}
 				}
+				continue
+			}
+
+			// If no claim found
+			clientAllocs, err := verifregState.GetAllocations(d.Client)
+			if err != nil {
+				return fmt.Errorf("getting allocations for client %s: %w", d.Client, err)
+			}
+			alloc, ok := clientAllocs[d.AllocationID]
+			if !ok {
+				// The allocation does not exist anymore.
+				// Either it was claimed and then claim was cleaned up after TermMax
+				// or allocation expired before it could be claimed and was cleaned up
+				// Deal should be cleaned up in either case
+				err = p.pd.RemoveDealForPiece(p.ctx, d.PieceCID, d.ID.String())
+				if err != nil {
+					// Don't return if cleaning up a deal results in error. Try them all.
+					log.Errorf("cleaning up direct deal %s for piece %s: %s", d.ID.String(), d.PieceCID, err.Error())
+				}
+				continue
+			}
+
+			// If claim is found then we should check if it is expired. If expired, clean up the deal.
+			if alloc.Expiration < head.Height() {
```

Comment:
these are the same logic, maybe just lump them in together, `if !ok || alloc.Expiration < head.Height() {`